### PR TITLE
Packet++/s7comm

### PR DESCRIPTION
**Pull request by wivien19**
- [x] S7comm
- [x] COTP
- [x] TPKT

I implemented S7comm, which based on COTP and TPKT. I also wrote test.

### CODE REVIEW

File: Packet++/CMakeLists.txt [hunk 1]
```diff
@@ -55,7 +55,9 @@ add_library(
   src/VlanLayer.cpp
   src/VxlanLayer.cpp
   src/WakeOnLanLayer.cpp
-  # Force hash-library pcapng to be link fully static
+  src/TpktLayer.cpp
+  src/CotpLayer.cpp
+  src/S7commLayer.cpp
```
Comment (clementperon) [inline Packet++/CMakeLists.txt:60]: Sort alphabetically

File: Packet++/CMakeLists.txt [hunk 2]
```diff
@@ -115,7 +117,11 @@ set(public_headers
     header/UdpLayer.h
     header/VlanLayer.h
     header/VxlanLayer.h
-    header/WakeOnLanLayer.h)
+    header/WakeOnLanLayer.h
+    header/TpktLayer.h
+    header/CotpLayer.h
+    header/S7commLayer.h)
```
Comment (clementperon) [inline Packet++/CMakeLists.txt:123]: sort alphabetically

File: Packet++/CMakeLists.txt [hunk 3]
```diff
@@ -51,16 +53,17 @@ add_library(
   src/TelnetLayer.cpp
   src/TextBasedProtocol.cpp
   src/TLVData.cpp
+  src/TpktLayer.cpp
   src/UdpLayer.cpp
   src/VlanLayer.cpp
   src/VxlanLayer.cpp
   src/WakeOnLanLayer.cpp
-  # Force hash-library pcapng to be link fully static
```
Comment (clementperon) [inline Packet++/CMakeLists.txt:58]: This comment has been involuntary sorted and should be next to $<TARGET_OBJECTS:hash-library>

File: Packet++/header/ProtocolType.h [hunk 4]
```diff
@@ -51,7 +51,18 @@ namespace pcpp
 	 * UDP protocol
 	 */
 	const ProtocolType UDP = 0x10;
-
+	/**
+	 * TPKT protocol
+	 */
+	const ProtocolType TPKT = 0x05;
```
Comment (clementperon) [inline Packet++/header/ProtocolType.h:57]: this should also be kept sorted regarding the value

File: Tests/Packet++Test/PacketExamples/tpkt.dat [hunk 5]
```diff
@@ -0,0 +1 @@
...720302502053f6bd650cac5f40f198cb37cdf8bae994704589b712dd1466b5813270dded4f3370000c890400000000000096430005cb7c02dcffcc03920000000b00140084000400009801000000000000000001800002120040010200010232003801000000ffffffffffffffffffffffffffffffff000100011164656661756c742d6d7270646f6d61696e00000217000801000014001400040240002401000a78643135303035363731000000e0dca079984400000a4900deffffff000a4900010251001c0101003f000000000000000000000000000000000000000000000000040000ec010000000000000000018001020f00a4010000000001800108706f72742d30303104000008706f72742d30303205706e2d696f0000000000001b1b05b6c9000008706f72742d3030310569706330350000000000000e8cec72b6000008706f72742d3030310c706c63786233313538623338000000000000000e8cd0f9cb000008706f72742d3030310a706c637862316430656400000000286336844ce9000000100000000000000000000000010000000000010215001401010000ffffffffffffffffffffffffffffffff0251001c01010008008c60610064acfe0000000000000000000000000000000004000070010000000000000000018002020f0028010000000001800208706f72742d30303200000000000000000000000000000000020000000000010215001401010000ffffffffffffffffffffffffffffffff0251001c01010008000000000000000000000000000000000000000000000000920000001600140000000000000072030000
```
Comment (seladb) [inline Tests/Packet++Test/PacketExamples/tpkt.dat:1]: Can you please add a `pcap` file that includes this packet? It's easier for viewing it in Wireshark etc.

File: Packet++/src/TpktLayer.cpp [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+		if (CotpLayer::isCotpPort(cotpType)) {
+			m_NextLayer = CotpLayer::parseCotpLayer(payload, payloadLen, this, m_Packet);
+		} else
+			m_NextLayer = new PayloadLayer(payload, payloadLen, this, m_Packet);
+
+	}
+
+	TpktLayer *TpktLayer::parseTpktLayer(uint8_t *data, size_t dataLen, Layer *prevLayer, Packet *packet) {
+		if (dataLen < sizeof(tpkthdr))
+			return NULL;
```
Comment (seladb) [inline Packet++/src/TpktLayer.cpp:73]: PcapPlusPlus moved to C++11 so we can use `nullptr` instead of `NULL`

File: Packet++/src/TpktLayer.cpp [hunk 7]
```diff
@@ -0,0 +1,94 @@
+#include "EndianPortable.h"
+#include "TpktLayer.h"
+#include "../header/TpktLayer.h"
```
Comment (seladb) [inline Packet++/src/TpktLayer.cpp:3]: nit: you can just do:
```cpp
include "TpktLayer.h"
```

File: Packet++/src/TpktLayer.cpp [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+	std::string TpktLayer::toString() const {
+		std::ostringstream vrsnStream;
+		vrsnStream << getVrsn();
+		std::ostringstream reservedStream;
+		reservedStream << getReserved();
+		std::ostringstream lengthStream;
+		lengthStream << getLength();
+
+		return "Tpkt Layer, vrsn: " + vrsnStream.str() +
```
Comment (seladb) [inline Packet++/src/TpktLayer.cpp:51]: nit: let's use `version` instead of `vrsn`?

**Review by seladb — COMMENTED** [review]
@wivien19 thank you so much for working on this, very much appreciated! 🙏 🙏 

I only reviewed `TpktLayer` and made some comments. Some of them are also relevant to the other layers as well. Please address them and then I'll review the rest of the PR.

If you prefer, you can also open 3 different PRs - one for each layer. That way it'll be easier to review

**Comment by seladb** [comment]
@wivien19 are you planning to continue working on this PR?

**Comment by wivien19** [comment]
> @wivien19 are you planning to continue working on this PR?

Yes, I plan to continue. This will take a few  more days.

**Comment by seladb** [comment]
@wivien19 there are conflicts with `dev` branch, can you please fix them before we further review the PR?

**Comment by seladb** [comment]
@wivien19 there are still conflicts with the `dev` branch. 

I suggest you split this PR into 3 smaller PRs: one for TPKT, one for COTP, and one for S7comm. 

Maybe you can start with opening the TPKT PR, get it reviewed and merged and then open the other 2 PRs? It'll be easier for us to review and for you to track and fix the comments.

Let me know what you think,

**Pull request closed without merging**